server/api/users/users.test.ts

```diff
@@ -37,19 +37,66 @@ describe('auth router', () => {
   });
 
   describe('GET /', () => {
-    it('expects empty array if no users exist', async () => {
+    it('expects empty array if no users exist on the homepage', async () => {
       const res = await server.exec.get('/api/users/');
-      expect(res.body).toStrictEqual([]);
+      expect(res.body.users).toStrictEqual([]);
     });
 
-    it('returns all users', async () => {
-      const userOne = await createUser();
-      const userTwo = await createUser();
+    it('expect invalid page request to return invalid page', async () => {
+      const res = await server.exec.get('/api/users/?page=-1');
+      expect(res.status).toBe(400);
+      expect(res.body.error).toEqual('Invalid page number');
+    });
+
+    it('expect page request beyond page limit to be out of range', async () => {
+      const res = await server.exec.get('/api/users/?page=100');
+      expect(res.status).toBe(400);
+      expect(res.body.error).toEqual('Page out of range');
+    });
+
+    it('returns first 20 users when it hits the homepage', async () => {
+
+      const users: User[] = [];
+
+      for(let i = 0; i < 25; i++){
+        const user = await createUser();
+        users.push(user);
+      }
+
+      const limit = 20;
+      const count = await User.count();
+      const totalPages = Math.ceil(count / limit);
 
       const res = await server.exec.get('/api/users/');
       expect(res.status).toBe(200);
-      expect(res.body[0]).toEqual(getExpectedUserObject(userOne));
-      expect(res.body[1]).toEqual(getExpectedUserObject(userTwo));
+      expect(res.body.users.length).toEqual(limit);
+      expect(res.body.users[0]).toEqual(getExpectedUserObject(users[0]));
+      expect(res.body.users[19]).toEqual(getExpectedUserObject(users[19]));
+      expect(res.body.page).toEqual(1);
+      expect(res.body.totalPages).toEqual(totalPages);
```

Comment:
This is testing a lot of stuff and it's being re-tested below. It should probably be broken up into smaller tests and the one below can then be simplified too.

I'd probably break it out into these tests:
- API endpoint exists
  - Simple test, hit the endpoint, check for status `200`. No need to create users.
- Number of items returned (this test's main goal)
  - Create 25 users, hit endpoint, check length of `res.body.users`.
- Sort order of items
  - Create 5 users, hit endpoint, `expect(_.map(res.body.users)).toEqual(_.map(users, 'id'), 'id')`
- Validity of metadata returned
  - Break into multiple tests to check 0 users, 20 users, 21 users
  - In the future, when we add customizable limits, we can reduce the counts for better test performance

Reply:
Is "Sort order of items" the correct name here? Wouldn't something like "created users match expected output of users" be more accurate?

Reply:
Why do we need the metadata response test?

I plan on checking if the API endpoint exists that the length of the array, page, and totalPages matches in the 20-user test.

I'll have the new commits shortly and it may make it more clear.

Reply:
Here's what I've done to satisfy your requests

1. API endpoint exists: I've created an API endpoint exists test
2. Number of items returned: I kept returns an array of 20 users, and returns last 5 users when it hits second page. I think checking the max length and second page is important using `res.body.users.length`
3. Sort order of items: was not created. I'm not entirely sure how to implement this request using the example you provided.
4. Validity of metadata in response: was not created. I have a comment above addressing it.